File: BaekJoon/Java/14427_수열과쿼리15_PriorityQueue.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

// 수열과쿼리15 14427 PriorityQueue
// O( N logN + M logN )


public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;

        int n = Integer.parseInt(br.readLine());
        int[] arr = new int[n]; // 원본 배열

        st = new StringTokenizer(br.readLine());
        // 최소 힙, { 값, 인덱스 }
        PriorityQueue<int[]> pq = new PriorityQueue<>((a, b) ->
                (a[0] == b[0]) ? a[1] - b[1] : a[0] - b[0]);
      
        for (int i = 0; i < n; ++i) {
            arr[i] = Integer.parseInt(st.nextToken());
            pq.offer(new int[]{arr[i], i + 1}); // 1-based index
        }

        int m = Integer.parseInt(br.readLine());

        while (m-- > 0) {
            st = new StringTokenizer(br.readLine());
            int cmd = Integer.parseInt(st.nextToken());

            // 현재 최소값의 인덱스 출력
            if (cmd == 2) {
                // 최소힙의 최상단 요소가 현재 배열의 값과
                // 일치하지 않으면 poll
                // 최신 상태의 { 값, 인덱스 }를 보장하기 위함
                while (!pq.isEmpty() && pq.peek()[0] != arr[pq.peek()[1] - 1]) {
                    pq.poll();
                }
                System.out.println(pq.peek()[1]);
            }
            // 배열&힙 값 업데이트
            else {
                int idx = Integer.parseInt(st.nextToken());
                int val = Integer.parseInt(st.nextToken());
                arr[idx - 1] = val;
                pq.offer(new int[]{val, idx});
            }
        }
    }
}

File: BaekJoon/14427_수열과쿼리15_PriorityQueue.cpp
#include <iostream>
#include <queue>
#include <vector>

using namespace std;
using pii = pair<int, int>;

// 수열과쿼리15 14427 PriorityQueue
// O( N logN + M logN )


class MinHeap {
private:
    // 최소 힙, { 값, 인덱스 }
    priority_queue<pii, vector<pii>, greater<>> pq;
    // 원본 배열
    vector<int> arr;
public:
    MinHeap(const vector<int>& v) : arr(v) 
    {
        for (int i = 0; i < arr.size(); ++i)
        {
            pq.emplace(arr[i], i + 1); // 1-based index
        }
    }

    // 배열&최소힙 값 업데이트
    void update(int idx, int val)
    {
        arr[idx] = val;
        pq.emplace(val, idx + 1);
    }

    // 현재 최소값의 인덱스 반환
    // Lazy Propagation (느린 갱신)
    int query()
    {
        // 최소 힙의 최상단 요소가 현재 배열의 값과
        // 일치하지 않으면 pop
        // 최신 상태의 { 값, 인덱스 }을 보장하기 위함
        while (pq.top().first != arr[pq.top().second - 1])
        {
            pq.pop();
        }
        return pq.top().second;
    }
};

int main(void)
{
    ios::sync_with_stdio(false);
    cin.tie(NULL);
    cout.tie(NULL);

    int n, m;
    cin >> n;

    vector<int> arr(n);
    for (int& x : arr)
    {
        cin >> x;
    }

    MinHeap minHeap(arr);

    cin >> m;
    
    while (m--)
    {
        int cmd;
        cin >> cmd;

        if (cmd == 2)
        {
            cout << minHeap.query() << "\n";
        }
        else
        {
            int idx, val;
            cin >> idx >> val;
            minHeap.update(idx - 1, val);
        }
    }


    return 0;
}
